New Plugin: betterNotifications

ctih1 opened the pull request:
Improves discord's desktop notifications, allowing further customization, more features (such as the ability to preview attachments), and overall nicer visuals

| ![Example screenshot on Windows](https://github.com/user-attachments/assets/421ab856-5377-41d5-985b-cae9e06b8188) |
| - | 
| Example screenshot of a notification containing an attachment |

---


<details>
<summary>The config used for the above screenshot </summary>

![image](https://github.com/user-attachments/assets/29a301f9-e54b-424e-a925-e55f311a52e1) 

</details>

## Compatibility

| Feature | Platforms |
| -------- | --------- |
Notification title customization | Windows, macOS, Linux |
| Notification body customization | Windows, macOS, Linux |
| Attachment preview | Windows, Linux* |
| Support for text formatting (aka **bolding**, *italic* and [links](https://github.com/Vendicated/Vencord/pull/3430) | Linux** |
| Avatar customization (cropping) | Windows, Linux |
| Attribute text | Windows |
| Headers (aka notification grouping | Windows |

*: An attachment will replace profile pictures on Linux
**: Only certain systems support this

### Note on macOS:
I haven't gotten Native notifications to work on macOS yet, and I'm not sure if it's even possible due to code signing issues. So instead of hijacking discord's notification implementation, we simply replace the variables that are given to the notification. This method still allows customizing the notification title and body. 
 
## Future plans
| Plan | status |
| ------ | --------- |
| Support for conditional rendering of variables (for example: `[nickname if mentioned]` | ✔️  |
| Support for conditional rendering of text (for example: `["mentioned" if mentioned]` | ✔️  |
| Support for conditional rendering of variables in notification titles | ✔️  |
| Improve the look of the settings page | ✔️  |
| Replace notification when a new one is received* | Windows |
| Auto hide notifications on spam | ✔️ |
| Inline reply** | Linux, macOS*** |
| Clicking the notification focuses Discord | ✔️ |
| Preview multiple attachments at once | :heavy_check_mark: |

\* Notifications automatically stack or disappear on Linux
** Support for inline replies on Windows would require the use of WinRT, which would require a dependency
*** Discord's native notifications already support replying on macOS

Here are some more examples:

<details>
<summary>Windows 10</summary>


<img width="376" height="323" alt="image" src="https://github.com/user-attachments/assets/603a7d16-6f72-4b31-a21c-57b8bc98690e" />

Incoming call:
<img width="384" height="239" alt="image" src="https://github.com/user-attachments/assets/ae55735c-5b1b-4be5-8964-caf49c587b40" /> |

Spoiler image:
<img width="378" height="324" alt="image" src="https://github.com/user-attachments/assets/392d3dbc-f7e6-4f00-a1be-060aa88bc2c3" />

Quick reactions:
<img width="372" height="194" alt="image" src="https://github.com/user-attachments/assets/bc2c122e-b5b1-461d-ab1d-44a9163e3c33" />

Multiple images:
<img width="383" height="373" alt="image" src="https://github.com/user-attachments/assets/7af975b5-2f89-454b-a454-eb25d6b480c2" />
</details>

<details>
<summary>Windows 11</summary>

<img width="610" height="319" alt="image" src="https://github.com/user-attachments/assets/39ed32f5-ed42-49c5-97ee-e12f9ac508e9" />

<img width="601" height="631" alt="image" src="https://github.com/user-attachments/assets/d72c27ce-0260-4fa0-892e-4a72e8e858a2" />


</details>

<details>
<summary>macOS</summary>



<img width="776" height="342" alt="image" src="https://github.com/user-attachments/assets/6585dfb8-f3d2-4abf-9d48-f6aea29e43f4" />

<img width="1174" height="638" alt="image" src="https://github.com/user-attachments/assets/70fc478d-eb32-4b61-bb73-977221158231" />


</details>

<details>
<summary>KDE (Manjaro)</summary>

<img width="351" height="141" alt="image" src="https://github.com/user-attachments/assets/9089d670-e1aa-4168-9e0d-7cf5f8006cc5" />

<img width="336" height="227" alt="image" src="https://github.com/user-attachments/assets/f3cea825-60b7-4a3b-82e6-c5cc5274f52c" />

</details>

<details>
<summary>Ubuntu</summary>

<img width="522" height="119" alt="image" src="https://github.com/user-attachments/assets/509d5715-09fa-4927-a7fc-1a4e8f9c2bcb" />

<img width="542" height="127" alt="image" src="https://github.com/user-attachments/assets/5e4bea8b-31dd-4de9-80ca-afd0cd4233bd" />

</details>

Respond to this PR if you have any feature requests!

ctih1 committed:
Added betterNotifications plugin and added myself to Devs constant

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
Removed unused setting

ctih1 committed:
Merge branch 'main' of https://github.com/ctih1/Vencord

ctih1 committed:
Fixed Windows notification header action

ctih1 committed:
specified activationType in notification Header

ctih1 committed:
added new patching strategy that makes the plugin work on macOS

ctih1 committed:
fixed macos detection

ctih1 committed:
added setting to ignore bots, and fixed notification click actions in direct messages

ctih1 committed:
Fixed bot detection since the bot variable seems to be always false?

ctih1 committed:
removed unecessary second-instance trigger

ctih1 committed:
fixed notification title in dms on macOS

ctih1 committed:
added cache options into settings

ctih1 committed:
Fixed xml error

ctih1 committed:
added variables for guild information

ctih1 committed:
removed debug notification

ctih1 committed:
removed more debug messages...

ctih1 committed:
improved format settings

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
removed unecessary newlines from components/variablestring.tsx, and fixeda spelling error (paranthesis -> parenthesis)

LazyStability commented:
Hi, concern Linux support, there is a standard  for notifications by the free desktop foundation. Most features your pr add, like attachment preview are specified. See this:
https://specifications.freedesktop.org/notification-spec/latest
Thank you for working on this!

ctih1 commented:
> Hi, concern Linux support, there is a standard for notifications by the free desktop foundation. Most features your pr add, like attachment preview are specified. See this: https://specifications.freedesktop.org/notification-spec/latest Thank you for working on this!

Thanks! I'll certainly look into it

ctih1 committed:
Merge branch 'main' into main

LazyStability commented:
> and the latter being that attachments seem to be capped to a very low resolution (at least on Ubuntu). 

This is by design, is it a problem? I just tested it and it looked fine for me. The relevant paragraph

>  Images may be placed in the notification, but this should be done with caution. The image should never exceed 200x100, but this should be thought of as a maximum size. Images should always have alternative text provided through the alt="..." attribute.
Image data cannot be embedded in the message itself. Images referenced must always be local files.
Implementations  are not required to support images. 

I have big concerns about the latter part however, if only local files are allowed, doesn't that kill the preview feature completely?

ctih1 commented:
> I have big concerns about the latter part however, if only local files are allowed, doesn't that kill the preview feature completely?

The attachments are already stored to the machine's %temp% folder on windows, as I found that using URLs was very buggy, and often caused the images to not load properly in the notification

LazyStability commented:
Well ok, that would work. How big is the image preview on windows or mac if I may ask?

LazyStability commented:
Ah ok, the attachment is the night sky on top of the notification, I thought that was noise mb. Yeah I don't think that will work on linux at all. I got confused by the cat. 

|Feature| Works on linux|
| ------------- | ------------- |
|Notification title customization 	| ✅|
Notification body customization 	|✅|
Attachment preview| ❌|
Avatar customization |✅|
Attribute text |❌|
Headers (aka notification grouping)|❌|

Not sure about the last one, I'd assume its the example author in the first image of the pr description.

cootshk committed:
chore: format

cootshk committed:
Fun fact: Kangaroo was one of Ven's old discord names

https://discord.com/channels/1015060230222131221/1015063227299811479/1379312919937945601

LazyStability commented:
Something like this works,  I wouldn't claim feature parity though.
![cat-notifications](https://github.com/user-attachments/assets/5eec18a3-1231-43a2-ac1c-ff054bb791fc)

cootshk committed:
fix: only check if you're on mac if the plugin is enabled

cootshk committed:
chore: format components

cootshk committed:
feat: allow prefixing for channels but not DMs

cootshk committed:
add myself as an author

cootshk committed:
fix: preview properly shows prefixes now

Covkie committed:
feat: linux support with notify-send

fix: proper avatar fetching
fix: prettier

ctih1 commented:
> Not sure about the last one, I'd assume its the `example author` in the first image of the pr description.

I'm pretty sure the headers are a Windows specific feature (?). They're used to group multiple notifications together in the action center, but also add a visual indicator to the message. I'll try to take a screenshot of this behaviour once I get my laptop working

ctih1 committed:
Merge pull request #2 from Covkie/pr/3430

feat: linux support with notify-send

cootshk committed:
fix merge conflicts

ctih1 committed:
Improved setting descriptions, added a setting to show spoiler images, and improved how attachments are processed

cootshk committed:
Revert "add myself as an author"

This reverts commit f7b8c74aeaa42ab0a3cafe5848b56370d743f0a5.

cootshk committed:
fix another merge conflict

ctih1 committed:
added variable hints for notification settings

ctih1 committed:
added vencord header

ctih1 committed:
removed unecesary listener

ctih1 committed:
fixed reactions causing the plugin to error. TODO: Implement a better way to show reactions.

ctih1 committed:
Merge branch 'betterNotifications' of https://github.com/cootshk/Vencord into cootshk-betterNotifications

ctih1 committed:
Merge branch 'cootshk-betterNotifications'

ctih1 committed:
merged

ctih1 committed:
Improved username- and channel prefix settings, as well as attribuite text settings.

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
added README and added credits for other contributors

ctih1 committed:
changed wording in credits

ctih1 committed:
added warning about Author.bot

ctih1 committed:
fixed ExampleStrings component not saving settings

ctih1 committed:
changed note to deprecation warning

ctih1 committed:
added more info to blocked notifications

ctih1 committed:
added support for fancies call notifications

ctih1 committed:
fixed hardcoded channel prefix in header, cleaned up and reorganized code

ctih1 committed:
fixed user prefix not applying

cootshk committed:
I use a capital C in my name wherever I can

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
split up index.tsx into different files

ctih1 committed:
Merge pull request #3 from cootshk/patch-1

ctih1 committed:
Merge branch 'main' of https://github.com/ctih1/Vencord

Covkie committed:
fixed avatar / attachments in notifications on linux

ctih1 committed:
added back file:// prefix

ctih1 committed:
fixed 'Missing argument for -h' from notify-send on linux

ctih1 committed:
Fixed XML injection vulnerability in generateXML

Covkie committed:
Round pfps linux

Covkie committed:
only do fs image rounding on linux

ctih1 committed:
added basic styling for linux notifications

ctih1 committed:
fixed crashing on startup

ctih1 committed:
fixed wrong parser typing

ctih1 committed:
fixed invalid parser call in notification.ts

ctih1 committed:
added more debugging for formatting

ctih1 committed:
removed invalid character!

ctih1 committed:
made formatting work on linux, and made links work

ctih1 committed:
Merge pull request #4 from Covkie/feat/betterNotifications/linux-round-pfps

linux round pfps

ctih1 committed:
Merge branch 'main' into dev

ctih1 committed:
added warning to markup settings

ctih1 committed:
replaced custom parser with a built in one

Covkie committed:
renderer platform checking

ctih1 committed:
added options for attachment fitting on notifications

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
added support for header customization, and a guildtag variable

Covkie committed:
hide platform specific options & x-kde-urls

ctih1 committed:
Merge pull request #5 from Covkie/pr/3430

ctih1 committed:
merge and added functionality for blurring attachments

ctih1 committed:
switched to function arg avatarUrl

ctih1 committed:
Merge pull request #6 from ctih1/dev

merge into main

ctih1 commented:
> Is there anything stopping the plugin from being approved? I don't want it to be merged right now, but I'd rather get any issues out of the way before I implement new ones 

Bump
Going on a vacation next week,so tomorrow (Sunday) is the only day where I can do bugfixes / feature requests

ctih1 committed:
Added ability to preview multiple attachments at once

ctih1 committed:
made loop breka when attachment limit is reached

ctih1 committed:
added note to attachmentfit select and made fillInBlank default

ctih1 committed:
fixed spoiler blur being incosistent due to different image sizes

ctih1 committed:
changed comment on blurImage to reflect new default

ctih1 committed:
changed to using array.at to avoid notifications erroring out wthout attachments

ctih1 committed:
Fixed attachment detection

ctih1 committed:
removed unecessary # in call notifications

ctih1 committed:
added support for quick reactions

ctih1 committed:
added testing for getting notification id

ctih1 committed:
added a listener

ctih1 committed:
added inline reply field

ctih1 committed:
changed way how dbus monitor is listened to

ctih1 committed:
changed function order

ctih1 committed:
changed order again

ctih1 committed:
added logic for scanning

ctih1 committed:
added more logging

ctih1 committed:
confirming data is being recieved

ctih1 committed:
split by newlines

ctih1 committed:
added more logging

ctih1 committed:
Improved reply detection

ctih1 committed:
improved notification reply handling

ctih1 committed:
added note and fixed invalid args

ctih1 committed:
added more logging to initial listener

ctih1 committed:
moved reply logic

ctih1 committed:
replying now works.

ctih1 committed:
testing out quick reactions on linux

ctih1 committed:
changed for testing

ctih1 committed:
changed how quickreactions are passsed

ctih1 committed:
added more logging

ctih1 committed:
logging

ctih1 committed:
added more loggin

ctih1 committed:
ignoring comma do something better later

ctih1 committed:
Made linux switch work

ctih1 committed:
improved quick reaction settings

ctih1 committed:
Merge pull request #7 from ctih1/linux-input

Linux input

ctih1 committed:
Merge branch 'Vendicated:dev' into dev

ctih1 committed:
Merge pull request #8 from ctih1/dev

Dev

ctih1 committed:
Merge branch 'Vendicated:main' into main

ctih1 committed:
added automatic notification removal on windows

ctih1 committed:
Fixed use of deprecated elements, added spam filtering, and changed to using Discord's native notification sound. (note: the implementation for that was extremely janky as I just removed the return statement, allowing discord to continue it's original notification implementation. I haven't seen any problems with this though.

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
Fixed notification click not focusing discord

ctih1 committed:
fixed visibility quirk and maybe fixed focus on Linux?

ctih1 committed:
Improved replacement logic by skipping discord's native notification instead of returning the entire function. If the user is using variable replacement, discord's native notification still gets triggered

ctih1 committed:
Changed macos warning period and removed reduntant commented out return

ctih1 committed:
Perhaps fixed notification duplication on linux?

ctih1 committed:
found a better work around for custom notifications

ctih1 committed:
added support for conditional rendering of variables

ctih1 committed:
Added support for strings instead of variables in notificatiosn

ctih1 commented:
One of the biggest things I want to improve is the settings page. It's not particularly beautiful, but I don't really know what I should do with it. **Any recommendations?**

![Part of the current settings page](https://github.com/user-attachments/assets/c6931f9d-40f1-41ef-aa42-e0b180f35908)

ctih1 committed:
Fixed variables in Replacement strategy and changed settings order a little

ctih1 committed:
Fixed getChannelInfoFromTitle returning unknown if server uses special characters in channel name)

ctih1 committed:
added a quick and easy way to create conditional statements. TODO: rearrange them since the modal is too small to contain them all

ctih1 committed:
changed spoiler things to be automatically enabled

ctih1 commented:
Here are some more examples:

<details>
<summary>Windows 10</summary>


<img width="376" height="323" alt="image" src="https://github.com/user-attachments/assets/603a7d16-6f72-4b31-a21c-57b8bc98690e" />

Incoming call:
<img width="384" height="239" alt="image" src="https://github.com/user-attachments/assets/ae55735c-5b1b-4be5-8964-caf49c587b40" /> |

Spoiler image:
<img width="378" height="324" alt="image" src="https://github.com/user-attachments/assets/392d3dbc-f7e6-4f00-a1be-060aa88bc2c3" />

Quick reactions:
<img width="372" height="194" alt="image" src="https://github.com/user-attachments/assets/bc2c122e-b5b1-461d-ab1d-44a9163e3c33" />

Multiple images:
<img width="383" height="373" alt="image" src="https://github.com/user-attachments/assets/7af975b5-2f89-454b-a454-eb25d6b480c2" />
</details>

I'll try to get some more examples in the upcoming days, but my only device with Windows 11 is my school laptop, which has some borked permissions, causing `pnpm inject` to fail

ctih1 committed:
Changed how variables are treated, added support for them in more notification parts, rewrote some bad code, and fixed issues with multiple variables

ctih1 committed:
Fixed notifications erroring out if avatarUrl is undefined

ctih1 committed:
fixed plugin erroring out when a reaction notification comes

sadan4 commented:
Do you know if this is even possible on Linux or do you only have Windows and Mac devices?

ctih1 commented:
I've tested this on a Linux virtual machine before, and it worked fine. Of course, some of the features (mainly inline reply) only work with certain notification daemons.

ctih1 commented:
This is basically the worst-case scenario, where the notification server itself just doesn't support any of the modern features. I've only seen this on a fresh install of Kubuntu. I've also tried this plugin on Ubuntu and Manjaro (with KDE Plasma), and it worked well there. I'll try to see if I have any old screenshots

<img width="1277" height="786" alt="image" src="https://github.com/user-attachments/assets/04086f8c-b6f1-4ea6-93b5-5258a1603e51" />

<img width="345" height="104" alt="image" src="https://github.com/user-attachments/assets/00e82e87-a8d1-4b26-b91d-cbb981a8c4b6" />

ctih1 commented:
Updated examples with Windows 11

ctih1 committed:
Fixed up some of the settings, and added a custom value for maximum attachment size

ctih1 committed:
Fixed forceFocus() not working like half the time because Windows..

ctih1 committed:
added webp support

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
Merge branch 'main' into main

ctih1 committed:
Merge branch 'main' into main
